doc/axiom.md

```diff
@@ -66,51 +68,121 @@ The characters after the underscore can be written as any word character; period
+```sparql
+"2022-01-01T10:10:10Z"^^xsd:dateTime
+```
 
 ### Numbers
 
 Examples: `0`, `-2`, `3.14`
 
-Numbers cover both integers and doubles, which are represented as integer and double literals in SPARQL, respectively. Neither are transformed during SPARQL translation beyond stringification.
+Numbers cover both integers and doubles, which are represented as integer and double literals in SPARQL, respectively. Neither are transformed during SPARQL translation beyond stringification. Note that the datatype IRI will vary depending on whether it is a double or integer, as well the underlying representation (e.g. Clojure integers will have the datatype IRI `xsd:long` as they are Java long values by default).
 
 ### Booleans
 
 Examples: `true` and `false`
 
 Booleans are not transformed during SPARQL translation beyond stringification.
 
-### Simple strings
+### Strings
 
 Examples: `"Hello World!"`, `"你好世界"`, `"cat: \\\"meow\\\""`, `"foo\\nbar"`
 
-In Flint, string literals can contain any characters **except** unescaped line breaks, carriage returns, backslashes, or double quotes; this is in order to prevent SPARQL injection attacks. (Therefore strings like `"cat: \"meow\"` and `"foo\nbar"` are not allowed.) Strings are not transformed during SPARQL translation.
+String literals can contain any characters **except** unescaped line breaks, carriage returns, backslashes, or double quotes; this is in order to prevent SPARQL injection attacks. (Therefore strings like `"cat: \"meow\"` and `"foo\nbar"` are not allowed.) Strings are not transformed during SPARQL translation.
 
-### Language-tagged strings
+### Language Maps
 
 Examples: `{:en "Hello World!"}`, `{:zh "你好世界"}`
 
-In Flint, strings with language tags are represented by a map between **one** language tag keyword and the string literal (which follows the same restrictions as simple strings).
+Strings with language tags are represented by a map between **one** language tag keyword and the string literal (which follows the same restrictions as simple strings). Note that this is the only literal that cannot have a datatype IRI appended, even if `:force-iris?` is `true`.
 
-### dateTime timestamps
+### Timestamps
 
 Examples: `#inst "2022-01-01T10:10:10Z"`
 
-In Flint, dateTime timestamps are any values for which `inst?` is `true`. During SPARQL translation, an IRI denoting the datatype is added as a suffix (since dateTime timestamps, unlike other literals, do not have a non-suffixed representation in SPARQL):
-```sparql
-"2022-01-01T10:10:10Z"^^<http://www.w3.org/2001/XMLSchema#dateTime>
+Timestamps are any values for which `inst?` is `true`, i.e. an instance of one of the following:
+- `java.time.Instant` (Clojure)
+- `java.util.Date` (Clojure)
+- `java.sql.Date` (Clojure)
+- `java.sql.Time` (Clojure)
```

Comment:
The reason why I extended the `Literal` protocol to all these different types is because they all satisfy `inst?` in Clojure, as you can see with [java.util.Date](https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L6783),   and [java.time.Instant](https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L6799). I guess I could remove the `java.sql` classes since they extend `java.util.Date`, but I want to keep `java.util.Date` equivalent to `xsd:dateTime`, especially since that is actually the default class an `#inst` literal gets coerced to.

Reply:
Yes, I understand the rationale; but at least for some of the work I do, the distinction between `xsd:date` and `xsd:dateTime` in particular is useful.  In grafter we chose to map platform and native clojure types to equivalent xsd types.

A big issue with `java.util.Date` is that almost all of the class has been deprecated since java 1.1; so it's a poor choice for many uses.  In my experience `xsd:dateTime`'s are much more common than `xsd:date`'s so I thought it made most sense to use the most useful types for them (i.e. `java.time.OffsetDateTime` etc); and map `java.util.Date` to `xsd:date`; or maybe don't just map it at all to discourage people from using it in the first place.

In grafter we tended to try and avoid coercion and instead focus on accurately representing the types.  This means there's a responsibility/burden on the developer, for example if they want a `java.util.Date` to be an `xsd:dateTime` they need to coerce it themselves for it not to be truncated.  However it's useful to use builtin types to make things relatively seamless, and it works quite well once you know what to expect.

I don't claim grafter is perfect in this regard; there were a few mistakes along the way, which were hard to correct later.

Reply:
There are reasons that I would want to keep the mapping of `java.util.Date` to `xsd:dateTime`:

- As you probably know, `java.util.Date` instances include time information (granted the class name obscures this fact). Therefore it wouldn't make sense to automatically coerce it to `xsd:date` and lose that time information.

- As I mentioned, `#inst` literals map to `java.util.Date` instances. I want users to be able to include literals in Flint representations and have those literals map to `xsd:dateTime` representations, without any additional conversion process:
```clojure
{:prefixes {:foo "<http://foo.com#>"
 :select ['?x]
 :where [['?x :foo/time #inst "2022-08-08T18:00:00Z"]]}
```
(As a side note, since you mentioned `java.time.OffsetDateTime`, it might be a good idea to have them and other `java.time.Temporal`-implementing classes implementing the `Literal` protocol regardless.)

Reply:
(Added additional notes and tests for timestamps in https://github.com/yetanalytics/flint/pull/27/commits/60fe82f798159bad0bfa7a1ad393f9563ac59aaf)

Reply:
Yes it's not by no means an unreasonable decision.  However it's worth mentioning that `#inst` and `java.util.Date` still lose information on the timezone.  e.g. the `Z` in your example will be silently lost; as would the timezone with `#inst "2022-08-08T18:00:00+03:00"`.

I guess this is arguably a clojure problem more than anything, I think due to clojure officially supporting java 1.5 when the `clojure.instant` namespace was written back in 2012.  In some ways it's a shame they didn't hold back on including until they could declare a minimum support for java 1.8, and adopt the `java.time` stuff more deeply.

> As a side note, since you mentioned java.time.OffsetDateTime  ... On the other hand you did recommend to be conservative with types, so we might leave those implementations to the library user.

Yes, by being conservative with types I mainly meant being cautious mapping too many things to the same type, when you could potentially use those types to represent other things such as differentiating between `xsd:date` and `xsd:dateTime`.  However, in the case of `java.time.OffsetDateTime` I'd support that being mapped to `xsd:dateTime`.

I think the main thing from my perspective is that you're not assuming `java.util.Date` and clojure inst's are good enough for all usecases as they don't support timezones.  So I'd be delighted to see `java.time.OffsetDateTime` mapped to that.

Time is complicated and there are a bunch of cases here as `xsd:dateTime`'s support serialising datetimes both with or without timezones; which are treated differently in the JVM.  Similarly there are other types such as `xsd:gYear`'s and periods which you may want to consider supporting at some point.  If you choose to do that I'd argue for considering those options in a different PR though.

Reply:
I decided to implement the rest of the timestamp-like `java.time.Temporal` classes in https://github.com/yetanalytics/flint/pull/27/commits/5309dc68c14d1816a7e89173f4cf4d1d1e24472f. That covers `ZonedDateTime`, `OffsetDateTime`, `LocalDateTime`, and their date- and time-only variants. (I also thought about including duration and `gXXX` types but decided that's probably outside the scope of this PR.)